[12.0][ADD] stock_picking_mail_incorrect_qty: Send an e-mail when a stock picking is incomplete

carmenbianca opened the pull request:
Task: https://gestion.coopiteasy.be/web#id=7745&view_type=form&model=project.task&action=479

Formerly https://github.com/coopiteasy/cie-custom/pull/40

carmenbianca committed:
[ADD] stock_picking_mail_incorrect_qty: Send an e-mail when a stock picking is incomplete

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

robinkeunen commented on stock_picking_mail_incorrect_qty/data/mail_template.xml:
```diff
@@ -0,0 +1,52 @@
+                ${ctx.get('company').email|safe}
+            </field>
+            <field name="subject">Incorrect Delivery (${object.name})</field>
+            <field name="partner_to">${ctx.get('company').email|safe}</field>
```
On bees database, it will send the email at `membre@bees-coop.be`. It's probably a good default value but it may be a good idea to say in `usage.rst` to configure this template with the good recipient.

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,96 @@
+        for pick in self:
+            pick.too_many_received_move_ids = pick._filtered_moves().filtered(
+                lambda move: move.quantity_done > move.product_qty
+            )
```
It's better practice to put compute methods depending on the same fields together. It saves quite a few loops. + these computations are closely related. Something like:

```py
    @api.multi
    @api.depends("move_lines")
    def _compute_move_discrepancy(self):
        for pick in self:
            filtered_moves = pick._filtered_moves()
            pick.zero_received_move_ids = filtered_moves.filtered(
                lambda move: move.quantity_done == 0 and move.product_qty
            )
            pick.too_few_received_move_ids = filtered_moves.filtered(
                lambda move: move.quantity_done < move.product_qty
            )
            pick.too_many_received_move_ids = filtered_moves.filtered(
                lambda move: move.quantity_done > move.product_qty
            )
```

\+ `pick` -> `picking`
\+ adapt `compute` argument in field definitions.

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,96 @@
+    @api.multi
+    def _compute_form_view_url(self):
+        for pick in self:
+            pick.form_view_url = (
+                "{}/web#id={}&model=stock.picking&view_type=form".format(
+                    pick.env["ir.config_parameter"].sudo().get_param("web.base.url"),
+                    pick.id,
+                )
+            )
```
I think it's better to call the action opening the picking form view.

carmenbianca replied:
I was going to ask about this. How do you call that action? Furthermore, is a computed field the correct place to do this? I need access to the URL in the template, and I don't have access to a lot of stuff inside of the template.

Is it possible to pass additional context into the template, so that I don't need to rely on a computed field at all?

robinkeunen replied:
Mmmh, if it's the email, calling the action won't work :/ So building the url manually is probably the best way to go. But you don't have to use a computed field, you can call a method from the object passed to the template.

carmenbianca replied:
Moved this into a function called `form_view_url()`. It has to be public for the mail template.

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,96 @@
+    @api.multi
+    def _filtered_moves(self):
+        return self.mapped("move_lines").filtered(
+            lambda move: move.state
```
`move` is actually a `move.line`.

carmenbianca replied:
It isn't :) `move_lines` = `move`. `move_line_ids` = `move.line`. It's horribly confusing.

robinkeunen replied:
damn, it is, you asked that question earlier --'

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,96 @@
+        return True
+
+    @api.multi
+    def _filtered_moves(self):
```
You could be more descriptive on what you're filtering here I think. Or maybe to it directly in the `_compute_move_discrepancy` body since it's now called once and is a one liner.

carmenbianca commented:
@robinkeunen Added your suggestions.

A few questions also:

- auto_delete is currently set to True in the template. Is this a correct default?
- Is there anything I need to do for translations?
- There's a chance that the code may not run correctly. Summarised below…

Upstream `stock` has this code:

```python
    def action_done(self):
        # TDE FIXME: remove decorator when migration the remaining
        todo_moves = self.mapped('move_lines').filtered(lambda self: self.state in ['draft', 'waiting', 'partially_available', 'assigned', 'confirmed'])
        # Check if there are ops not linked to moves yet
        for pick in self:
            # # Link existing moves or add moves when no one is related
            for ops in pick.move_line_ids.filtered(lambda x: not x.move_id):
                # Search move with this product
                moves = pick.move_lines.filtered(lambda x: x.product_id == ops.product_id)
                moves = sorted(moves, key=lambda m: m.quantity_done < m.product_qty, reverse=True)
                if moves:
                    ops.move_id = moves[0].id
                else:
                    new_move = self.env['stock.move'].create({
                                                    'name': _('New Move:') + ops.product_id.display_name,
                                                    'product_id': ops.product_id.id,
                                                    'product_uom_qty': ops.qty_done,
                                                    'product_uom': ops.product_uom_id.id,
                                                    'location_id': pick.location_id.id,
                                                    'location_dest_id': pick.location_dest_id.id,
                                                    'picking_id': pick.id,
                                                    'picking_type_id': pick.picking_type_id.id,
                                                   })
                    ops.move_id = new_move.id
                    new_move = new_move._action_confirm()
                    todo_moves |= new_move
                    #'qty_done': ops.qty_done})
        todo_moves._action_done()
        self.write({'date_done': fields.Datetime.now()})
        return True
```

The entire for-loop that handles a corner case (move.lines unlinked to moves, but linked to stock.picking) is not executed by our module. There may therefore be moves that are invisible to us. Ideally upstream would have SoC'd this and extracted the for-loop into a separate function `create_moves_for_unlinked_move_lines()`, but that's not the case. If we want to handle this case, we have to:

- Copy the for-loop verbatim.
- Submit a patch for `create_moves_for_unlinked_move_lines()` to upstream, but I don't believe they're taking patches for Odoo 12.
- Put the patch in OCB (the upstream one, or our own fork).

Thoughts?

robinkeunen commented:
Better late than never

### auto_delete is currently set to True in the template. Is this a correct default?

```py
class MailMail(models.Model):
    """ Model holding RFC2822 email messages to send. This model also provides
        facilities to queue and send new email messages.  """
    _name = 'mail.mail'
    ...
    auto_delete = fields.Boolean(
        'Auto Delete',
        help="Permanently delete this email after sending it, to save space")
```

Yes, we don't need to keep the history of these warning emails.

### Is there anything I need to do for translations?

No, translations are managed by analysts and customers. From time to time, they ask us to bring back the terms they translated in the GUI to the VCS.

### There's a chance that the code may not run correctly. Summarised below…

> Copy the for-loop verbatim.

Unfortunatly, this is the way to go because :

> Submit a patch for create_moves_for_unlinked_move_lines() to upstream, but I don't believe they're taking patches for Odoo 12.

Odoo will not indeed patch odoo 12, and

> Put the patch in OCB (the upstream one, or our own fork).

OCB will only take fixes that were submitted to odoo/odoo but refused. This is not really a fix you're suggesting.

And we don't patch OCB ourselves, we prefer to keep open branches and merge them with gitaggregate.
It seems fragile that this module should depend on a patched version of our OCB.

But do put your PR comment in the comments of your `def action_done`

carmenbianca committed:
[IMP] stock_picking_mail_incorrect_qty: Include upstream code to prevent bug

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[REF] stock_picking_mail_incorrect_qty: Use stock.move.line instead of stock.move

For some reason---on some databases---there is a discrepancy between the
two when manually adding items to the picking. stock.move.line would
contain the new item, but stock.move wouldn't. Using stock.move.line
here seems like a decent enough work-around.

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca commented:
@robinkeunen Would love a review on the refactor commit. (Never mind the ugly formatting; it's `prettier`'s new look apparently).

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,130 @@
+
+    @api.multi
+    def action_done(self):
+        # The following code between the 'fmt' tags is copied almost-verbatim
+        # from the stock module. It handles a corner case of move.lines being
+        # unlinked to moves, but linked to stock.picking. It is copied here
+        # verbatim because not including it would (potentially) result in some
+        # moves being skipped over.
```
I don't understand why you need to copy this code here since you call super later ?

carmenbianca replied:
The order I want:

- Copied code snippet from stock_picking.action_done(). This fixes discrepancies as described.
- Send e-mail
- The remainder of stock_picking.action_done(). This function 'destroys' the information that is needed for composing the e-mail (e.g., empty moves are cancelled, quantities are split and adjusted, etc etc etc).

If I do the following order:

- Send e-mail
- All of stock_picking.action_done()

… then the e-mailing functionality may not have access to the fixed discrepancies.

_However_, because I refactored this to use stock.move.lines instead of stock.moves\*, there may be a chance that we are now no longer affected by the possible discrepancies. So I guess this can be removed?

\*: I don't particularly like this change. The module worked fine without the change … just not on bees.test.

robinkeunen replied:
> The order I want:
>  ...

Ok I get it. Since it's in addons, I guess we can commit this. Your explanation about order of execution would be welcome in the comments as well.

> _However_, because I refactored this to use stock.move.lines instead of stock.moves*, there may be a chance that we are now no longer affected by the possible discrepancies. So I guess this can be removed?

Not sure about that, I see _rabbit holes_ everywhere in stocks now, we'll need to test.

> *: I don't particularly like this change. The module worked fine without the change … just not on bees.test.

Did you check in the shell the modules used on bees.test ? I remember of a module to invoice of product from picking lines that were not in the sale order. I cannot find the name of the module right now and scrum is calling.

Thanks for your investigations

carmenbianca replied:
@robinkeunen I added the above explanation as a comment to the code, and I have reverted the change from stock.move to stock.move.line. Turns out I was using the wrong variable for what I really wanted. Why it displayed different behaviour on bees compared to not-bees, I may never know. I couldn't find any module with `odoo_repl` that altered any relevant code, and I crawled up the code tree quite a bit.

robinkeunen commented on stock_picking_mail_incorrect_qty/models/stock_picking.py:
```diff
@@ -0,0 +1,130 @@
+    def _notify_incorrect_delivery(self):
+        """Send a notification e-mail about the incorrect delivery."""
+        for picking in self:
+            if not any(
+                (
+                    picking.zero_received_move_line_ids,
+                    picking.zero_expected_move_line_ids,
+                    picking.too_few_received_move_line_ids,
+                    picking.too_many_received_move_line_ids,
+                )
+            ):
+                continue
+
+            self.env.ref(
+                "stock_picking_mail_incorrect_qty.mail_template_incorrect_delivery"
+            ).send_mail(picking.id)
```
I prefer keeping positive conditions where possible :

```py
if any(...):
    self.env.ref(...).send_mail(picking.id)
```

carmenbianca replied:
Makes sense :)

robinkeunen reviewed (commented):
I have questions about the copied code bit.

Also, I'd name the module `stock_picking_mail_unexpected_incorrect_qty`

carmenbianca commented:
> Also, I'd name the module `stock_picking_mail_unexpected_incorrect_qty`

I think unexpected is superfluous here, if we generously interpret 'I received a pizza when I asked for salad' to be an incorrect delivery.

carmenbianca commented:
TODO: Tidy up the commits. Remove the reverted commit entirely. Also run pre-commit again.

carmenbianca committed:
Revert "[REF] stock_picking_mail_incorrect_qty: Use stock.move.line instead of stock.move"

This reverts commit 5a64ef34212aeb1643d41d975ca6ca5356e5194e.

carmenbianca committed:
[FIX] stock_picking_mail_incorrect_qty: Run pre-commit

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] stock_picking_mail_incorrect_qty: Do not send mails for POS pickings

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] stock_picking_mail_incorrect_qty: SCRLfs → SC

Signed-off-by: Carmen Bianca Bakker <carmen@coopiteasy.be>
